a little bit speed-up for full node syncing

garyyu opened the pull request:
This PR can save > 60% time on full node syncing.

Today I need a full node syncing and find it always wait 5 seconds for every new block_sync requesting, but only 1st second or first 2 seconds has real downloading.

This change will detect the fact that there's no more new blocks downloaded in 200ms, and then allow a new block_sync requesting in this case.

garyyu commented:
Some test logs here. As you can see, the average interval is less than 2 seconds, compared to 5 seconds before this PR.
```
Sep 02 23:24:55.064 DEBG body_sync: 17/65 blocks received, and no more in 200ms
Sep 02 23:24:56.977 DEBG body_sync: 15/48 blocks received, and no more in 200ms
Sep 02 23:24:58.599 DEBG body_sync: 10/33 blocks received, and no more in 200ms
Sep 02 23:25:00.175 DEBG body_sync: 9/23 blocks received, and no more in 200ms
Sep 02 23:25:01.757 DEBG body_sync: 5/14 blocks received, and no more in 200ms
Sep 02 23:25:03.083 DEBG body_sync: 3/9 blocks received, and no more in 200ms
Sep 02 23:25:04.447 DEBG body_sync: 1/5 blocks received, and no more in 200ms
Sep 02 23:25:05.759 DEBG body_sync: 2/4 blocks received, and no more in 200ms
Sep 02 23:25:20.151 DEBG body_sync: 12/100 blocks received, and no more in 200ms
Sep 02 23:25:22.201 DEBG body_sync: 34/100 blocks received, and no more in 200ms
Sep 02 23:25:24.061 DEBG body_sync: 33/100 blocks received, and no more in 200ms
Sep 02 23:25:25.960 DEBG body_sync: 30/86 blocks received, and no more in 200ms
Sep 02 23:25:27.775 DEBG body_sync: 24/58 blocks received, and no more in 200ms
Sep 02 23:25:29.397 DEBG body_sync: 16/38 blocks received, and no more in 200ms
Sep 02 23:25:30.880 DEBG body_sync: 5/21 blocks received, and no more in 200ms
Sep 02 23:25:32.474 DEBG body_sync: 11/47 blocks received, and no more in 200ms
Sep 02 23:25:34.592 DEBG body_sync: 19/48 blocks received, and no more in 200ms
Sep 02 23:25:36.639 DEBG body_sync: 27/100 blocks received, and no more in 200ms
Sep 02 23:25:38.507 DEBG body_sync: 33/100 blocks received, and no more in 200ms
Sep 02 23:25:40.349 DEBG body_sync: 23/72 blocks received, and no more in 200ms
Sep 02 23:25:42.184 DEBG body_sync: 19/54 blocks received, and no more in 200ms
Sep 02 23:25:43.727 DEBG body_sync: 8/35 blocks received, and no more in 200ms
Sep 02 23:25:45.637 DEBG body_sync: 30/84 blocks received, and no more in 200ms
Sep 02 23:25:47.466 DEBG body_sync: 19/54 blocks received, and no more in 200ms
Sep 02 23:25:49.084 DEBG body_sync: 13/35 blocks received, and no more in 200ms
Sep 02 23:25:50.346 DEBG body_sync: 3/22 blocks received, and no more in 200ms
Sep 02 23:25:52.060 DEBG body_sync: 13/61 blocks received, and no more in 200ms
Sep 02 23:25:53.703 DEBG body_sync: 14/47 blocks received, and no more in 200ms
Sep 02 23:25:55.293 DEBG body_sync: 13/32 blocks received, and no more in 200ms
Sep 02 23:25:57.331 DEBG body_sync: 33/100 blocks received, and no more in 200ms
Sep 02 23:25:59.088 DEBG body_sync: 31/86 blocks received, and no more in 200ms
Sep 02 23:26:00.913 DEBG body_sync: 17/57 blocks received, and no more in 200ms
Sep 02 23:26:02.577 DEBG body_sync: 18/49 blocks received, and no more in 200ms
Sep 02 23:26:04.158 DEBG body_sync: 10/29 blocks received, and no more in 200ms
Sep 02 23:26:06.184 DEBG body_sync: 29/94 blocks received, and no more in 200ms
Sep 02 23:26:07.822 DEBG body_sync: 24/65 blocks received, and no more in 200ms
Sep 02 23:26:09.343 DEBG body_sync: 9/41 blocks received, and no more in 200ms
Sep 02 23:26:10.881 DEBG body_sync: 9/32 blocks received, and no more in 200ms
Sep 02 23:26:12.411 DEBG body_sync: 7/23 blocks received, and no more in 200ms
Sep 02 23:26:14.186 DEBG body_sync: 7/16 blocks received, and no more in 200ms
Sep 02 23:26:15.578 DEBG body_sync: 3/9 blocks received, and no more in 200ms
Sep 02 23:26:16.815 DEBG body_sync: 2/6 blocks received, and no more in 200ms
Sep 02 23:26:23.075 DEBG body_sync: 1/4 blocks received, and no more in 200ms
Sep 02 23:26:29.276 DEBG body_sync: 1/3 blocks received, and no more in 200ms
Sep 02 23:26:30.581 DEBG body_sync: 1/2 blocks received, and no more in 200ms
Sep 02 23:26:34.351 DEBG body_sync: 43/99 blocks received, and no more in 200ms
Sep 02 23:26:36.177 DEBG body_sync: 36/100 blocks received, and no more in 200ms
Sep 02 23:26:37.959 DEBG body_sync: 27/100 blocks received, and no more in 200ms
Sep 02 23:26:39.565 DEBG body_sync: 24/82 blocks received, and no more in 200ms
Sep 02 23:26:41.698 DEBG body_sync: 19/57 blocks received, and no more in 200ms
Sep 02 23:26:43.300 DEBG body_sync: 13/38 blocks received, and no more in 200ms
Sep 02 23:26:44.922 DEBG body_sync: 8/36 blocks received, and no more in 200ms
Sep 02 23:26:46.807 DEBG body_sync: 22/67 blocks received, and no more in 200ms
Sep 02 23:26:48.658 DEBG body_sync: 21/69 blocks received, and no more in 200ms
Sep 02 23:26:50.463 DEBG body_sync: 17/48 blocks received, and no more in 200ms
Sep 02 23:26:52.009 DEBG body_sync: 11/31 blocks received, and no more in 200ms
Sep 02 23:26:53.591 DEBG body_sync: 6/26 blocks received, and no more in 200ms
Sep 02 23:26:55.600 DEBG body_sync: 28/95 blocks received, and no more in 200ms
Sep 02 23:26:57.270 DEBG body_sync: 17/69 blocks received, and no more in 200ms
Sep 02 23:26:59.026 DEBG body_sync: 16/52 blocks received, and no more in 200ms
Sep 02 23:27:00.574 DEBG body_sync: 11/35 blocks received, and no more in 200ms
Sep 02 23:27:02.491 DEBG body_sync: 23/81 blocks received, and no more in 200ms
Sep 02 23:27:04.328 DEBG body_sync: 25/86 blocks received, and no more in 200ms
Sep 02 23:27:05.880 DEBG body_sync: 15/61 blocks received, and no more in 200ms
Sep 02 23:27:07.590 DEBG body_sync: 13/46 blocks received, and no more in 200ms
Sep 02 23:27:09.464 DEBG body_sync: 25/75 blocks received, and no more in 200ms
Sep 02 23:27:11.110 DEBG body_sync: 17/50 blocks received, and no more in 200ms
Sep 02 23:27:12.481 DEBG body_sync: 6/32 blocks received, and no more in 200ms
Sep 02 23:27:14.243 DEBG body_sync: 10/26 blocks received, and no more in 200ms
Sep 02 23:27:16.198 DEBG body_sync: 28/94 blocks received, and no more in 200ms
Sep 02 23:27:17.824 DEBG body_sync: 20/66 blocks received, and no more in 200ms
Sep 02 23:27:19.564 DEBG body_sync: 14/46 blocks received, and no more in 200ms
Sep 02 23:27:21.297 DEBG body_sync: 12/32 blocks received, and no more in 200ms
Sep 02 23:27:23.085 DEBG body_sync: 9/20 blocks received, and no more in 200ms
```

ignopeverell commented:
Seems that this could be simplified. You're tracking every single block hash that sync requests and check if one of these has been received, and the 200ms gap. You could just track whether the head has been update in the last 200ms, right?

garyyu commented:
No, we could not track the head, because it's possible that received blocks are just put into orphans queue, without any update on the chain head.

garyyu commented:
Perhaps we could check both head and orphans queue?

garyyu commented:
Updated for the simplified version to avoid tracking every single block hash that sync requests.


And **an interesting thing** found after adding `len_evicted()` for `OrphanBlockPool`:  
**the evicting will never happen!**
because in `body_sync()` function:

```
	let block_count = cmp::min(
		cmp::min(100, peer_count * 10),
		chain::MAX_ORPHAN_SIZE.saturating_sub(chain.orphans_len()) + 1,
	);
```
This limit the maximum possible orphans, it's impossible to get more orphans than MAX_ORPHAN_SIZE.

If this is the expected behavior, I think all the following code can be removed from `add()` function of `OrphanBlockPool`:

```
		if orphans.len() > MAX_ORPHAN_SIZE {
			let old_len = orphans.len();

			// evict too old
			orphans.retain(|_, ref mut x| {
				x.added.elapsed() < Duration::from_secs(MAX_ORPHAN_AGE_SECS)
			});
			// evict too far ahead
			let mut heights = height_idx.keys().cloned().collect::<Vec<u64>>();
			heights.sort_unstable();
			for h in heights.iter().rev() {
				if let Some(hs) = height_idx.remove(h) {
					for h in hs {
						let _ = orphans.remove(&h);
					}
				}
				if orphans.len() < MAX_ORPHAN_SIZE {
					break;
				}
			}
			// cleanup index
			height_idx.retain(|_, ref mut xs| xs.iter().any(|x| orphans.contains_key(&x)));

			self.evicted
				.fetch_add(old_len - orphans.len(), Ordering::Relaxed);
		}
```

@ignopeverell   Can we remove it?
If yes, then I can remove the new added `len_evicted()` and related code, for more simple code:)

antiochp commented:
We should be careful here with edge cases.

A node may only be peered with one or two other peers.
We only pull historical blocks from "archive nodes" and we have no special peering logic (may understanding here may be wrong or out of date) to ensure we are peered with a healthy number of these (may even be 0 archival node peers).

So we can easily get into situations where we are asking for historical blocks and it takes longer than 200ms to receive any blocks.

garyyu commented:
@antiochp  Is that the case of 1 second  timeout?   Look https://github.com/mimblewimble/grin/blob/f5bab7c80148f1bf528de8e0cb19d7f3de5906e8/servers/src/grin/sync.rs#L251-L260

> A node may only be peered with one or two other peers.
Network connectivity may have high latency and/or just be really slow.

In this case, node could need a few days to complete full sync, or even never complete. As you said, this is the typical **edge** case.  For full syncing, I think we can only focus on common cases, and can ignore such kind of edge case. 

If the node's network is so bad, why we need make it happy to a full sync?  let the good nodes serve the Grin network well:)

garyyu commented:
@ignopeverell  There's a question above which need your help, thanks:)

ignopeverell commented:
I do think latency is an issue and over internet a round trip can easily take multiple seconds. However I think we can cater to both needs with just a small addition: make sure we got at least one block back. As you're tracking the download of each block with good granularity now, we could make sure that:

1. Request for more blocks get sent.
2. Wait for a first block to come in or 5 sec, whichever comes first (so 5 secs is more of a timeout in case all requests got lost somehow).
3. Once we got at least one block, do the 200ms check.

Does that help?

garyyu commented:
Thanks, and please also look that question at 2 days ago:  `@ignopeverell Can we remove it? `

ignopeverell commented:
Oh, sorry about that, entirely missed it. The check on add is still required however as orphans will also be used after sync, during normal block broadcast. With enforcing a max, we could get tons of orphaned block bloating memory.

garyyu commented:
Thanks  @ignopeverell   it makes sense for block broadcast, so check on add is indeed necessary.

And regarding to following proposals:
> Request for more blocks get sent.

But this need to change `MAX_ORPHAN_SIZE` also, currently it's `200`.  I'm not sure you agree to enlarge it to for example `400`, do you?

> Wait for a first block to come in or 5 sec, whichever comes first (so 5 secs is more of a timeout in case all requests got lost somehow).

This is exactly what I did in this PR :)  but I used 1 sec.  Now I change to 5 secs according to your proposal.

> Once we got at least one block, do the 200ms check.

OK.

garyyu commented:
Any more comments on this PR? otherwise it's ready to merge.

ignopeverell commented:
Sorry for the delay, all good for me, thanks!
